internal/upgradecheck/http.go

```diff
@@ -107,38 +105,40 @@ func checkForUpgrades(log logr.Logger, versionString string, backoff wait.Backof
 	// i.e., the process hits the cap for time or the number of steps
 	// The anonymous function here sets certain preexisting variables (res, err)
 	// which are then used by the surrounding `checkForUpgrades` function as part of the return
+	var bodyBytes []byte
+	var status int
+
 	if err == nil {
 		_ = wait.ExponentialBackoff(
 			backoff,
 			func() (done bool, backoffErr error) {
-				// We can't close the body of this response in this block since we use it outside
-				// so ignore this linting error
-				res, err = client.Do(req) //nolint:bodyclose
-				// This is a very basic check, ignoring nuances around
-				// certain StatusCodes that should either prevent or impact retries
-				if err == nil && res.StatusCode == http.StatusOK {
-					return true, nil
-				}
+				var res *http.Response
+				res, err = client.Do(req)
+
 				if err == nil {
-					err = fmt.Errorf("received StatusCode %d", res.StatusCode)
+					defer res.Body.Close()
+					status = res.StatusCode
+
+					// This is a very basic check, ignoring nuances around
+					// certain StatusCodes that should either prevent or impact retries
+					if status == http.StatusOK {
+						bodyBytes, err = io.ReadAll(res.Body)
+						return true, nil
+					}
 				}
+
 				// Return false, nil to continue checking
 				return false, nil
 			})
 	}
-	// If the final value of err is nil and the final res.StatusCode is OK,
-	// we can go on with reading the body of the response
-	if err == nil && res.StatusCode == http.StatusOK {
-		defer res.Body.Close()
-		bodyBytes, err = ioutil.ReadAll(res.Body)
-	}
 
-	// TODO: Parse response and log info for user on potential upgrades
-	if err == nil {
-		return string(bodyBytes), nil
+	// We received responses, but none of them were 200 OK.
+	if err == nil && status != http.StatusOK {
+		err = fmt.Errorf("received StatusCode %d", status)
 	}
 
-	return "", err
+	// TODO: Parse response and log info for user on potential upgrades
+	return string(bodyBytes), err
```

Comment:
Just to talk it out, when we get here, our potential results are

(any of the retries calls successfully)
* a successful call: string(bodyBytes) is the response, err is nil
* a successful call with a ReadAll error: string(bodyBytes) is probably "", err is the ReadAll error

(all the retries fail to get a 200 response)
* a non-errored, non-200 call: string(bodyBytes) is the response of the non-200 call, err is a message we set on l 137 abt the non-200 status
* error: string(bodyBytes) is "", err is the error message

Reply:
Agreed. There are two assignments to `err` inside the backoff callback/closure and those will be returned when `ExponentialBackoff` returns. `bodyBytes` is set right before `ExponentialBackoff` returns and will contain whatever we got from the read.

It seemed a little odd ignoring the return value from `ExponentialBackoff`, but we're not interested in knowing if backoff expired or not. Both before and now, we're interested in the values from the HTTP call.